plugin/src/main/scala/chisel3/internal/plugin/SerializableModule.scala

```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.internal.plugin
+
+import scala.collection.mutable
+import scala.tools.nsc
+import scala.tools.nsc.{Global, Phase}
+import scala.tools.nsc.plugins.PluginComponent
+import scala.tools.nsc.symtab.Flags
+import scala.tools.nsc.transform.TypingTransformers
+
+// Create a object `SomeSerializableModuleGenerator` for `SomeSerializableModule extends SerializableModule[SomeSerializableModuleParameter]`:
+// object SomeSerializableModuleGenerator extends App {
+//   emitMlirbc(SerializableModuleGenerator(classOf[SomeSerializableModule], upickle.default.read[P](json)(implicitly[upickle.default.Read[P]])))
+// }
+class SerializableModuleGenerator(val global: Global, arguments: ChiselPluginArguments)
+    extends PluginComponent
+    with TypingTransformers
+    with ChiselOuterUtils {
+  import global._
+
+  val phaseName = "serializablemodulegeneratorphase"
+  val runsAfter: List[String] = List[String]("typer")
+  def newPhase(_prev: Phase): SerializableModuleGeneratorPhase = new SerializableModuleGeneratorPhase(_prev)
+
+  class SerializableModuleGeneratorPhase(prev: Phase) extends StdPhase(prev) {
+    override def name: String = phaseName
+    def apply(unit: CompilationUnit): Unit = {
+      if (ChiselPlugin.runComponent(global, arguments)(unit)) {
+        unit.body = new MyTypingTransformer(unit).transform(unit.body)
+      }
+    }
+  }
+
+  private class MyTypingTransformer(unit: CompilationUnit) extends TypingTransformer(unit) with ChiselInnerUtils {
+    override def transform(tree: Tree): Tree = tree match {
+      case packageDef @ PackageDef(pid, stats) =>
+        val newStats = stats.flatMap {
+          case classDef: ClassDef if isASerializableModule(classDef.symbol) && !classDef.mods.hasFlag(Flag.ABSTRACT) =>
+            val parameterTypeName = classDef.tparams.filter(t => isASerializableModuleParameter(t.symbol)).head.name.toTypeName
+            val moduleTypeName = classDef.name.toTypeName
+            val moduleTermName = classDef.name.toTermName
+            val generatorObject = q"""object ${TermName(moduleTermName.toString + "Generator")} extends App { _root_.chisel3.experimental.SerializableModuleGenerator(classOf[${moduleTypeName}], upickle.default.read[${parameterTypeName}](args.head)(implictly[upickle.default.Read[${parameterTypeName}]])) } """
```

Comment:
Please break this line up. It's also really important to check that this generated object doesn't collide with anything existing--find a way to check the names of symbols in the companion object when generating it--what we do if there's a collision is debatable, we could have some unification algorithm, we could just not generate it, or we could error.

Reply:
Sure! I'm planing migrate to a trait implementation. I have some release ddl these two days, I'll go back to it this weekend.